src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,738 @@
+{
+    // io_error has different functionality, it will be set to errno which returned by system
+    // functions (such as read()).
+    int io_error = 0;
+
+    // Internal function to appending into the buffer.
+    ssize_t put(const char *msg, size_t count) noexcept;
+
+    // Throw exception based on current state
+    void throw_exception_on(const int states);
+
+    public:
+    ostream() noexcept
+    {
+        buf = std::unique_ptr<streambuf>(new(std::nothrow) streambuf);
+    }
+
+    ostream(std::unique_ptr<streambuf> passed_buf) noexcept
+    {
+        buf = std::move(passed_buf);
+    }
+
+    ostream(const int tfd) noexcept
+    {
+        fd = tfd;
+        buf = std::unique_ptr<streambuf>(new(std::nothrow) streambuf);
+    }
+
+    ostream(const int tfd, std::unique_ptr<streambuf> passed_buf) noexcept
+    {
+        fd = tfd;
+        buf = std::move(passed_buf);
+    }
+
+    ostream(const ostream &) = delete;
+
+    void operator=(const ostream &) = delete;
+
+    ostream(const ostream &&) = delete;
+
+    void operator=(const ostream &&) = delete;
+
+    // Open file path which specified in path or as open(2) parameter with flags or modes.
+    //
+    // Note: O_WRONLY is hard-coded in internal call to POSIX open(2).
+    //
+    // Note: Opening an already opened stream is not allowed and caller must close the stream
+    // before open it again.
+    //
+    // Returns: true on success, false on failure.
+    //
+    // Non _nx variant may throw:
+    // dio::iostream_system_err on POSIX open(2) failure
+    bool open_nx(const char *path) noexcept;
+    bool open(const char *path);
+
+    bool open_nx(const char *path, const int flags) noexcept;
+    bool open(const char *path, const int flags);
+
+    bool open_nx(const char *path, const int flags, const mode_t mode) noexcept;
+    bool open(const char *path, const int flags, const mode_t mode);
+
+    // Flush the buffer and close the current file descriptor.
+    //
+    // Returns: true on success, false on failure cases.
+    //
+    // Non _nx variant may throw:
+    // dio::iostream_internal_err on Buffer failure (buffer did not allocated at construction time)
+    // dio::iostream_system_err on system I/O failure
+    bool close_nx() noexcept;
+    bool close();
+
+    // Returns: current bits/vars in stream based on io_states enum.
+    int current_state() noexcept;
+
+    // Returns: true when there is no set bits/vars otherwise false.
+    bool good() noexcept;
+
+    // Returns: true when buffer_failbit is set otherwise false.
+    bool buffer_failure() noexcept;
+
+    // Returns: 0 or errno based on io_error value.
+    int io_failure() noexcept;
+
+    // Resets failure bits/vars to false (or 0 about io_error).
+    // Note: buffer_failbit will not be reset.
+    void clear() noexcept;
+
+    // Flush buffer of stream.
+    //
+    // Returns: the size of buffer on full-flushed buffer, -1 on failure or not full-flushed buffer.
```

Comment:
The comment (or implementation) is wrong, it currently just returns the  size of the last successful write that was used to flush the buffer contents.

If it only distinguishes between successful and unsuccessful flush, it should return a `bool` not an `int`.